src/internal/analytics/components/analytics-funnel.tsx

```diff
@@ -0,0 +1,152 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import React, { useEffect, useRef, useState } from 'react';
+
+import {
+  FunnelStepContext,
+  FunnelSubStepContext,
+  FunnelContext,
+  FunnelContextValue,
+} from '../context/analytics-context';
+import { useFunnel, useFunnelStep } from '../hooks/use-funnel';
+import { useUniqueId } from '../../hooks/use-unique-id';
+import { useVisualRefresh } from '../../hooks/use-visual-mode';
+
+import { PACKAGE_VERSION } from '../../environment';
+
+import FunnelMetrics from '../';
+import { FunnelProps, FunnelStepProps } from '../interfaces';
+
+import { getFunnelNameSelector, getSubStepAllSelector, getSubStepNameSelector, getSubStepSelector } from '../selectors';
+
+export const FUNNEL_VERSION = '1.0';
+
+export const AnalyticsFunnel: React.FC<
+  Pick<FunnelProps, 'funnelType' | 'optionalStepNumbers' | 'totalFunnelSteps'>
+> = ({ children, ...props }) => {
+  const [funnelInteractionId, setFunnelInteractionId] = useState<string>();
+  const funnelResultRef = useRef<boolean>(false);
+  const isVisualRefresh = useVisualRefresh();
+
+  // This useEffect hook is run once on component mount to initiate the funnel analytics.
+  // It first calls the 'funnelStart' method from FunnelMetrics, providing all necessary details
+  // about the funnel, and receives a unique interaction id.
+  // This unique interaction id is then stored in the state for further use.
+  //
+  // On component unmount, it checks whether the funnel was successfully completed.
+  // Based on this, it either calls 'funnelComplete' or 'funnelCancelled' method from FunnelMetrics.
+  //
+  // The eslint-disable is required as we deliberately want this effect to run only once on mount and unmount,
+  // hence we do not provide any dependencies.
+  useEffect(() => {
+    const funnelInteractionId = FunnelMetrics.funnelStart({
+      funnelNameSelector: getFunnelNameSelector(),
+      optionalStepNumbers: props.optionalStepNumbers,
+      funnelType: props.funnelType,
+      totalFunnelSteps: props.totalFunnelSteps,
+      componentVersion: PACKAGE_VERSION,
+      componentTheme: isVisualRefresh ? 'vr' : 'classic',
+      funnelVersion: FUNNEL_VERSION,
+    });
+
+    setFunnelInteractionId(funnelInteractionId);
+
+    return () => {
+      // Send funnelComplete or funnelCancelled event after the current event loop. This ensures that this metric will
+      // appear last in the event stream.
+      setTimeout(() => {
```

Comment:
In the real world is more of a convenience factor where this will then be the last event of the funnel. I'm not sure about the side effects for unit tests. If its problem worth considering then we could definitely take this out for now. Thoughts?

Reply:
I am afraid of getting warnings similar to AWSUI-20146. Fixing this later may require breaking changes on Panorama integration so we better to have a plan prepared already. I will follow-up in DM about Panorama implementation details (we should enforce some constraints what this integration can't do, return promises for example)